### ISSUE TITLE
Research: apple/coremltools for model conversion

### ISSUE THREAD

**Issue by ryanmaclean** [ryanmaclean/vibecode-webgui#1137]
Deep research apple/coremltools (convert and optimize ML models for Core ML). Evaluate relevance to on-device inference and potential for local model deployment.

**Comment by ryanmaclean** [comment]
Research summary: apple/coremltools

What it is / capabilities
- Apple’s official Python toolkit for Core ML model conversion, editing, and validation.
- Supports conversion from TensorFlow 1.x/2.x, PyTorch, and non‑NN frameworks like scikit‑learn, XGBoost, and LibSVM.
- Unified conversion API (`coremltools.convert`) is the recommended path for TF/PyTorch.

Current install guidance (from upstream)
- GitHub README calls out “Install Version 8.3” via `pip install -U coremltools` and a 9.0 beta (`pip install coremltools==9.0b1`).
- The official docs still show install instructions for 7.1 in some guides; versioning appears to move quickly, so we should pin a version and note the mismatch in docs.

PyTorch conversion workflow (high level)
1) Capture graph via `torch.jit.trace` or `torch.export.export`.
2) Convert with `coremltools.convert`.
- `torch.jit.trace` is the stable path; `torch.export` is newer/beta (Core ML Tools 8.0+).
- Unified conversion API targets iOS 13/macOS 10.15+; older deployment targets require the older onnx-coreml path.

Fit for us
- Useful if we want local/offline model conversion for macOS packaging or to ship Core ML models in the app.
- Requires a Python-based build step; we’d need to decide on version pinning + CI environment (macOS runners if we want validation/predictions).

Suggested next steps
1) Decide target deployment versions (macOS/iOS) → drives conversion path.
2) Run a tiny PoC converting a representative PyTorch model and measure fidelity/perf.
3) If we need legacy targets, plan an onnx-coreml fallback.

**Comment by ryanmaclean** [comment]
## Research Findings: Apple Core ML Tools (coremltools)

### Overview

**coremltools** is Apple's official Python library for converting and optimizing ML models to Core ML format (.mlpackage, .mlmodel). It serves as the primary bridge between popular ML frameworks (TensorFlow, PyTorch, scikit-learn, XGBoost, LibSVM) and Apple's on-device ML runtime.

**Repository**: https://github.com/apple/coremltools  
**Documentation**: https://apple.github.io/coremltools/

---

### Key Capabilities

#### 1. Unified Conversion API
- Single `ct.convert()` interface for all frameworks
- Supports TensorFlow 1/2, PyTorch, scikit-learn, XGBoost
- Outputs to ML Program (modern) or Neural Network (legacy) formats

```python
import coremltools as ct

# PyTorch conversion
mlmodel = ct.convert(traced_model, 
                     inputs=[ct.TensorType(shape=(1, 3, 224, 224))],
                     minimum_deployment_target=ct.target.iOS18)
```

#### 2. Advanced Model Optimization
- **Palettization**: Weight clustering (2-8 bits) with lookup tables
- **Linear Quantization**: INT8/FP16 weight quantization
- **Pruning**: Magnitude-based and structured sparsity
- **Joint Compression**: Combine multiple techniques

```python
import coremltools.optimize as cto

# 6-bit palettization
config = cto.coreml.OptimizationConfig(
    global_config=cto.coreml.OpPalettizerConfig(nbits=6)
)
compressed = cto.coreml.palettize_weights(mlmodel, config)
```

#### 3. Hardware-Specific Optimization
- **Apple Neural Engine (ANE)** targeting for transformers
- Compute unit selection (CPU, GPU, ANE, or all)
- Automatic graph optimization and operator fusion

---

### Relevance to On-Device Inference

#### Strong Alignment with Project Goals:

| Feature | Benefit |
|---------|---------|
| **Privacy-First** | Local inference - no data leaves device |
| **Offline Capability** | 100% offline operation |
| **Low Latency** | Direct hardware acceleration (ANE, GPU, CPU) |
| **Power Efficiency** | Optimized for mobile power budgets |
| **No Server Costs** | Zero inference infrastructure required |

#### Hardware Acceleration Support:
- **Apple Silicon**: M1/M2/M3/M4 series (Mac, iPad)
- **iOS Devices**: iPhone/iPad with A12 Bionic and later
- **Neural Engine**: Dedicated ML silicon for up to 15x speedup
- **GPU**: Metal Performance Shaders for parallel compute

---

### Local Model Deployment Potential

#### Supported Model Types:
1. **Vision**: CNNs, Vision Transformers, segmentation models
2. **NLP**: Transformers, BERT, GPT-style models, LLMs (Llama, Mistral)
3. **Audio**: Speech recognition, audio classification
4. **Multimodal**: Stable Diffusion, CLIP embeddings
5. **Traditional ML**: XGBoost, scikit-learn models

#### Real-World Success Stories:
- **Llama 3.1 on-device**: Apple optimized 8B parameter models for iPhone (Nov 2024)
- **Stable Diffusion**: Full pipeline conversion for macOS/iOS
- **Mistral 7B**: Running on iOS devices via Hugging Face + Core ML
- **WWDC 2024**: Demonstrated Apple Intelligence on-device foundation models

#### Deployment Workflows:
```python
# Export for iOS deployment
mlmodel.save("model.mlpackage")

# Or with compute unit specification
model = ct.models.MLModel("model.mlpackage", 
                          compute_units=ct.ComputeUnit.ALL)
```

---

### Technical Architecture

```
PyTorch/TensorFlow → coremltools → Core ML Format (.mlpackage)
                                          ↓
                              ┌──────────────────────┐
                              │   Core ML Runtime    │
                              │  ┌────────────────┐  │
                              │  │ Apple Neural   │  │
                              │  │ Engine (ANE)   │  │
                              │  ├────────────────┤  │
                              │  │    GPU         │  │
                              │  │  (Metal)       │  │
                              │  ├────────────────┤  │
                              │  │    CPU         │  │
                              │  └────────────────┘  │
                              └──────────────────────┘
```

---

### Strengths

✅ **Mature ecosystem** - battle-tested at Apple scale  
✅ **Comprehensive compression** - industry-leading quantization/palettization  
✅ **Apple hardware optimization** - direct access to ANE  
✅ **Active development** - regular updates (v8.x as of 2024)  
✅ **Production-ready** - used by Apple Intelligence, third-party apps  
✅ **Good documentation** - extensive guides and examples  

---

### Limitations

⚠️ **Apple ecosystem only** - iOS, macOS, watchOS, tvOS  
⚠️ **Limited custom ops** - constrained operator support vs. PyTorch  
⚠️ **Conversion complexity** - some models require manual intervention  
⚠️ **No training** - inference-only (though supports updatable models for on-device fine-tuning)  
⚠️ **Model size limits** - very large models (>10GB) may face memory constraints  

---

### Recommendation

**HIGHLY RELEVANT** for vibecode-webgui if targeting Apple platforms.

**Use Cases:**
1. **Primary deployment path** for iOS/macOS local ML
2. **Model compression pipeline** to reduce model sizes 50-90%
3. **Offline-first AI features** in the app
4. **Privacy-sensitive inference** (health data, personal content)

**Integration Strategy:**
- Use coremltools as the conversion layer in CI/CD
- Store compressed .mlpackage files in the app bundle
- Implement fallback to CPU-only mode for older devices

**Alternative Consideration:**
If cross-platform support (Android, Web) is needed, consider TensorFlow Lite or ONNX Runtime Mobile as primary with coremltools as Apple-specific optimization.

---

### Next Steps

1. **Prototype conversion** - Test with a small PyTorch model from the project
2. **Benchmark performance** - Compare Core ML vs. other mobile runtimes
3. **Evaluate model compatibility** - Check if target models convert successfully
4. **Assess compression gains** - Measure size/accuracy tradeoffs
5. **Test on real devices** - Validate ANE utilization and battery impact
